codalab/lib/bundle_store.py

```diff
@@ -1,20 +1,109 @@
-'''
+"""
 BundleStore is a data store that stores actual bundle data. Bundles are logical
 folders within this data store. This class provides two main methods:
   get_location: return the location of the folder with the given data hash.
   upload: upload a local directory to the store and return its data hash.
-'''
-import errno
+"""
 import os
 import time
 import sys
-import uuid
-import tempfile
+
+from hash_ring import HashRing
 
 from codalab.lib import path_util, file_util, print_util, zip_util
-from codalab.common import UsageError
 
-class BundleStore(object):
+def require_disks(f):
+    """Decorator added to MulitDiskBundleStore methods that require a disk to
+    be added to the deployment for tasks to succeed. Prints a helpful error
+    message prompting the user to add a new disk.
+    """
+    def wrapper(*args, **kwargs):
+        self = args[0]
+        if self.ring.get_node("DOESN'T MATTER") is None:
+            print >> sys.stderr,"""
+Error: No disks available.
+To use MultiDiskBundleStore, you must add at least one disk. Try the following:
+
+    $ cl help add-disk
+"""
+            sys.exit(1)
+        else:
+            return f(*args, **kwargs)
+    return wrapper
+
+
+class BundleStoreCleanupMixin(object):
+    """A mixin for BundleStores that wish to support a cleanup and full cleanup operation.
+    """
+    def cleanup(self, model, data_hash, except_bundle_uuids, dry_run):
+        """
+        Cleanup a given model
+        """
+        pass
+
+    def full_cleanup(self, model, dry_run):
+        """
+        For each data hash in the store, check if it should be garbage collected and
+        delete its data if so. In addition, delete any old temporary files.
+        """
+        pass
+
+
+class BaseBundleStore(object):
+    """
+    BaseBundleStore defines the basic interface that all subclasses are *required* to implement. Concrete subtypes of
+    this class my introduce new functionality, but they must all support at least these interfaces.
+    """
+
+    def __init__(self):
+        """
+        Create and initialize a new instance of the bundle store.
+        """
+        self.initialize_store()
+
+    def initialize_store(self):
+        """
+        Initialize the bundle store with whatever structure is needed for use.
+        """
+        pass
+
+    def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove_sources):
+        """
+        Allow the client to create a bundle via uploading a file, directory, or Git repository.
+        """
+        pass
+
+    def get_bundle_location(self, data_hash):
+        """
+        Gets the location of the bundle with cryptographic hash digest data_hash. Returns the location in the method
+        that makes the most sense for the storage mechanism being used.
+        """
+        pass
+
+    def get_bundle_temp_location(self, identifer):
```

Comment:
Those are good questions. I feel like the get_bundle_location API is general enough to do something like ex. lookup in an S3 bucket, maybe the upload functionality could be factored out into a separate piece of functionally (ex. another mixin class called FileSystemUpload that allows uploading files to a bundle store.

In the current formulation though I think it's general enough, as upload_bundle doesn't assume anything about the storage format, all the options are for the bundle lookup done on disk. I feel like the API for uploads really could use improvement but it's not something I was looking to tackle in this PR.